docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+
+However we need to provide a way for users to specify which class of loadbalancers to join, and which particular loadbalancer (in the case of multiple loadbalancer per class). Note that we can't just say each class of loadbalancers has a single loadbalancer in it, and so picking a class boils down to picking a specific loadbalancer, because of cloud provider limitations. Nor can we say picking a specific loadbalancer picks the class, because a loadbalancer might have limited capacity (i.e don't force a user to pick 172.12, which might be maxed out, when they just want class=gce).
+
+The proposed long term solution for this problem is to reference a claim on an ip through the Ingress Path.
+
+### How does TLS work?
+
+Each IngressPoint specifies a single secret and a tls mode. Its upto the loadbalancer controller to handle the tls mode. For example, the gce controller will only support TLS termination for 1.1 because each IngressPoint gets a new loadbalancer, and GCE doesn't allow multiple certs per IP. With the current design a user can create the wrong type of secret for the wrong mode. They wouldn't know till they try creating the Ingress path. Validating a simple key:value secret is also hard.
+
+Some alternatives to the proposed tls mode handling:
+* Remove tlsMode and create a new type of secret(s).
+* Don't use a secret, embed the TLSConfig in the IngressPoint.
+
+Regardless of how we implement it, we should probably treat tlsmode as a request and not a directive. Even if the user provides a secret the admin may choose to not expose it, or have a policy that controls that.
+
+### What triggers the creation of a new loadbalancer controller, should this be exposed to users?
+
+In the current proposal, the kube-controller-manager is started up with a --loadbalancer-controllers flag that dictates which controllers to startup. This is obviously sub-optimal. A better solution would be --loadbalancer-controllers=gce-lbc-pod,haproxy-lbc-pod..., then this just boils down to creating some rcs. Taking this a step further, we could have a resource to allow dynamically adding more loadbalancer controllers to the cluster.
+
+### How does public DNS work?
+
+Most of this is TDB. With the current model, if someone specifies a hostname the loadbalancer controllers assume they know what they're doing. If there are 2 IngressPoints with the same hostname and url endpoints, one of them will win. This is just like the overlapping labels rc problem. Since this is ultimately a policy and security decision that admins make, having replaceable/plugin controllers with each one choosing a policy might suffice.
```

Comment:
Yes, and it also handles duplicates (oldest route wins, and overlapping
paths in the same namespace are ok, but not ok across namespaces).  I think
it's generally useful for a routelet that has a wildcard DNS name assigned
to it to always assign a generated name, and also optionally assign a host
name.

On Tue, Sep 8, 2015 at 11:42 AM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/loadbalancing.md
> <https://github.com/kubernetes/kubernetes/pull/12827#discussion_r38941504>
> :
>
> > +
> > +Each IngressPoint specifies a single secret and a tls mode. Its upto the loadbalancer controller to handle the tls mode. For example, the gce controller will only support TLS termination for 1.1 because each IngressPoint gets a new loadbalancer, and GCE doesn't allow multiple certs per IP. With the current design a user can create the wrong type of secret for the wrong mode. They wouldn't know till they try creating the Ingress path. Validating a simple key:value secret is also hard.
> > +
> > +Some alternatives to the proposed tls mode handling:
> > +* Remove tlsMode and create a new type of secret(s).
> > +* Don't use a secret, embed the TLSConfig in the IngressPoint.
> > +
> > +Regardless of how we implement it, we should probably treat tlsmode as a request and not a directive. Even if the user provides a secret the admin may choose to not expose it, or have a policy that controls that.
> > +
> > +### What triggers the creation of a new loadbalancer controller, should this be exposed to users?
> > +
> > +In the current proposal, the kube-controller-manager is started up with a --loadbalancer-controllers flag that dictates which controllers to startup. This is obviously sub-optimal. A better solution would be --loadbalancer-controllers=gce-lbc-pod,haproxy-lbc-pod..., then this just boils down to creating some rcs. Taking this a step further, we could have a resource to allow dynamically adding more loadbalancer controllers to the cluster.
> > +
> > +### How does public DNS work?
> > +
> > +Most of this is TDB. With the current model, if someone specifies a hostname the loadbalancer controllers assume they know what they're doing. If there are 2 IngressPoints with the same hostname and url endpoints, one of them will win. This is just like the overlapping labels rc problem. Since this is ultimately a policy and security decision that admins make, having replaceable/plugin controllers with each one choosing a policy might suffice.
>
> I'm starting to move the OpenShift router into the direction of each
> router functioning like a kubelet - updating the route with information
> about the DNS name the router chose to expose for the service. A user who
> wants a specific DNS name may have to live with not getting it, or getting
> a second one. Either way, the router will report that alongside the route.
>
> Correct me if this doesn't translate to: If you specify a name the
> routelet assumes you own it, otherwise it assumes you don't really care and
> makes a best effort.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12827/files#r38941504>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift